code/functions.py

```diff
@@ -0,0 +1,165 @@
+import numpy as np
+import holoviews as hv
+from holoviews import plotting
+from holoviews import Options, Store
+import pfaffian as pf
```

Comment:
`pfaffian` doesn't really belong to functions. It's used only in weeks 1-2. Also it only defines a single useful function, so if anything, it should be
```python
from pfaffian import pfaffian
```
(I also suggest to not abbreviate `pfaffian` to `pf`: it's not used all too often).